In [7]:
import pandas as pd

In [8]:
pd.set_option('display.max_columns', None)

In [9]:
# Iterating over 3 years of flight data to create a dataframe, 
# choose which column data to keep while also rearranging them
# and append it to a list containing all the data frames

years = ["2020","2021","2022"]
dataframes = []
cols = [
    "FlightDate",
    "Year", "Quarter", "Month", "DayofMonth", "DayOfWeek",
    "Airline",
    "Origin", "OriginCityName", "OriginState",
    "Dest", "DestCityName", "DestState",
    "Cancelled", "Diverted",
    "CRSDepTime", "DepTime", "DepDelay",
    "CRSArrTime", "ArrTime", "ArrDelay",
    "DepDel15", "ArrDel15",
    "AirTime", "CRSElapsedTime", "ActualElapsedTime",
    "Distance",
    "TaxiOut", "TaxiIn"
    ]

for year in years:
    df = pd.read_csv("Combined_Flights_" + year + ".csv")
    df = df[cols]
    dataframes.append(df)

In [10]:
df_flightdata = pd.concat(dataframes)

In [11]:
df_flightdata.shape

(15412586, 29)

Concatanated dataframes contains 446,965,994 data cells

In [12]:
df_flightdata

,FlightDate,Year,Quarter,Month,DayofMonth,DayOfWeek,Airline,Origin,OriginCityName,OriginState,Dest,DestCityName,DestState,Cancelled,Diverted,CRSDepTime,DepTime,DepDelay,CRSArrTime,ArrTime,ArrDelay,DepDel15,ArrDel15,AirTime,CRSElapsedTime,ActualElapsedTime,Distance,TaxiOut,TaxiIn
0,2020-09-01,2020,3,9,1,2,Comair Inc.,PHL,"Philadelphia, PA",PA,DAY,"Dayton, OH",OH,False,False,1905,1858.0,-7.0,2056,2034.0,-22.0,0.0,0.0,76.0,111.0,96.0,477.0,16.0,4.0
1,2020-09-02,2020,3,9,2,3,Comair Inc.,PHL,"Philadelphia, PA",PA,DAY,"Dayton, OH",OH,False,False,1905,1858.0,-7.0,2056,2027.0,-29.0,0.0,0.0,68.0,111.0,89.0,477.0,16.0,5.0
2,2020-09-03,2020,3,9,3,4,Comair Inc.,PHL,"Philadelphia, PA",PA,DAY,"Dayton, OH",OH,False,False,1905,1855.0,-10.0,2056,2122.0,26.0,0.0,1.0,77.0,111.0,147.0,477.0,65.0,5.0
3,2020-09-04,2020,3,9,4,5,Comair Inc.,PHL,"Philadelphia, PA",PA,DAY,"Dayton, OH",OH,False,False,1905,1857.0,-8.0,2056,2027.0,-29.0,0.0,0.0,73.0,111.0,90.0,477.0,13.0,4.0
4,2020-09-05,2020,3,9,5,6,Comair Inc.,PHL,"Philadelphia, PA",PA,DAY,"Dayton, OH",OH,False,False,1905,1856.0,-9.0,2056,2026.0,-30.0,0.0,0.0,72.0,111.0,90.0,477.0,14.0,4.0
...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...,...
4078313,2022-03-31,2022,1,3,31,4,Republic Airlines,MSY,"New Orleans, LA",LA,EWR,"Newark, NJ",NJ,False,True,1949,2014.0,25.0,2354,234.0,NaN,1.0,NaN,NaN,185.0,NaN,1167.0,17.0,32.0
4078314,2022-03-17,2022,1,3,17,4,Republic Airlines,CLT,"Charlotte, NC",NC,EWR,"Newark, NJ",NJ,True,False,1733,1817.0,44.0,1942,NaN,NaN,1.0,NaN,NaN,129.0,NaN,529.0,NaN,NaN
4078315,2022-03-08,2022,1,3,8,2,Republic Airlines,ALB,"Albany, NY",NY,ORD,"Chicago, IL",IL,False,False,1700,2318.0,378.0,1838,59.0,381.0,1.0,1.0,135.0,158.0,161.0,723.0,19.0,7.0
4078316,2022-03-25,2022,1,3,25,5,Republic Airlines,EWR,"Newark, NJ",NJ,PIT,"Pittsburgh, PA",PA,False,True,2129,2322.0,113.0,2255,939.0,NaN,1.0,NaN,NaN,86.0,NaN,319.0,25.0,6.0


In [53]:
df_flightdata["DepDelay"].min()

-131.0

In [26]:
# displaying the names of all the columns
df_flightdata.columns

Index(['FlightDate', 'Year', 'Quarter', 'Month', 'DayofMonth', 'DayOfWeek',
       'Airline', 'Origin', 'OriginCityName', 'OriginState', 'Dest',
       'DestCityName', 'DestState', 'Cancelled', 'Diverted', 'CRSDepTime',
       'DepTime', 'DepDelay', 'CRSArrTime', 'ArrTime', 'ArrDelay', 'DepDel15',
       'ArrDel15', 'AirTime', 'CRSElapsedTime', 'ActualElapsedTime',
       'Distance', 'TaxiOut', 'TaxiIn'],
      dtype='object')

In [27]:
# displaying number of rows
len(df_flightdata)

15412586

In [28]:
# checking the data type that each column holds
df_flightdata.dtypes

FlightDate            object
Year                   int64
Quarter                int64
Month                  int64
DayofMonth             int64
DayOfWeek              int64
Airline               object
Origin                object
OriginCityName        object
OriginState           object
Dest                  object
DestCityName          object
DestState             object
Cancelled               bool
Diverted                bool
CRSDepTime             int64
DepTime              float64
DepDelay             float64
CRSArrTime             int64
ArrTime              float64
ArrDelay             float64
DepDel15             float64
ArrDel15             float64
AirTime              float64
CRSElapsedTime       float64
ActualElapsedTime    float64
Distance             float64
TaxiOut              float64
TaxiIn               float64
dtype: object

In [31]:
# counting number of occurences of each airline
df_flightdata["Airline"].value_counts()

Airline
Southwest Airlines Co.                       2757841
Delta Air Lines Inc.                         1841213
American Airlines Inc.                       1802158
SkyWest Airlines Inc.                        1791136
United Air Lines Inc.                        1107063
Republic Airlines                             750579
Endeavor Air Inc.                             620948
Envoy Air                                     617944
Comair Inc.                                   551099
JetBlue Airways                               503658
Alaska Airlines Inc.                          456691
Spirit Air Lines                              455671
Mesa Airlines Inc.                            367688
Frontier Airlines Inc.                        314874
Allegiant Air                                 287874
Horizon Air                                   264011
Capital Cargo International                   224517
Air Wisconsin Airlines Corp                   174545
Commutair Aka Champlain Enterprises, I

In [49]:
df_flightdata["Airline"].value_counts().describe()

count    2.500000e+01
mean     6.165034e+05
std      7.074909e+05
min      5.202000e+03
25%      1.539860e+05
50%      3.676880e+05
75%      6.209480e+05
max      2.757841e+06
Name: count, dtype: float64

Southwest Airlines Co. had the highest number of flights, whereas Empire Airlines Inc. had the lowest amount.

In [32]:
# counting how many airlines exist in data
len(df_flightdata["Airline"].value_counts())

25

In [72]:
df_flightdata["Cancelled"].value_counts(normalize = True).mul(100).round(2)

Cancelled
False    96.53
True      3.47
Name: proportion, dtype: float64

About 3.47% of flights were cancelled, while 96.53% were not.

In [71]:
df_flightdata["Diverted"].value_counts(normalize = True).mul(100).round(2)

Diverted
False    99.78
True      0.22
Name: proportion, dtype: float64

About 0.22% of flights were diverted, while 99.78% where not.

In [45]:
# grouping flights by airline, calculating the average delay time, and display the results in descending order.
df_flightdata.groupby("Airline")["DepDelay"].mean().sort_values(ascending = False).round(2)

Airline
JetBlue Airways                              15.24
Commutair Aka Champlain Enterprises, Inc.    13.84
Allegiant Air                                13.30
Mesa Airlines Inc.                           11.31
Frontier Airlines Inc.                       10.85
Spirit Air Lines                             10.37
GoJet Airlines, LLC d/b/a United Express      9.96
American Airlines Inc.                        9.79
Southwest Airlines Co.                        9.54
United Air Lines Inc.                         8.64
SkyWest Airlines Inc.                         8.29
Trans States Airlines                         7.65
Air Wisconsin Airlines Corp                   7.45
Comair Inc.                                   7.15
Delta Air Lines Inc.                          6.13
Republic Airlines                             4.98
Envoy Air                                     4.93
Capital Cargo International                   3.94
Endeavor Air Inc.                             3.67
Horizon Air            

JetBlue Airways had the highest average delay time of about 15 minutes and 14 seconds.

ExpressJet Airlines Inc. had the lowest average delay time of about 0.01 seconds.

In [50]:
# displayng JetBlue Airways average delay times over the years
df_flightdata[df_flightdata["Airline"] == "JetBlue Airways"].groupby("Year")["DepDelay"].mean().round(2)

Year
2020     1.47
2021    16.58
2022    26.01
Name: DepDelay, dtype: float64

JetBlue Airways experienced a significant increase
in average departure delays from 2020 to 2022.

Average delays increased from approximately 1.47 minutes
in 2020 to 26.01 minutes in 2022, indicating worsening
operational performance following the COVID-19 pandemic period.

In [57]:
# filtering out smaller airports and then checking their average delay times

airport_counts = df_flightdata["Origin"].value_counts()
large_airports = airport_counts[airport_counts > 10000].index
df_filtered = df_flightdata[df_flightdata["Origin"].isin(large_airports)]
df_filtered.groupby("Origin")["DepDelay"].mean().sort_values(ascending=False).round(2)

Origin
ASE    22.39
EYW    14.49
HPN    14.39
EWR    13.65
JAC    12.74
       ...  
ANC     2.53
FAI     2.51
LIH     1.70
JNU     1.03
ITO     0.14
Name: DepDelay, Length: 161, dtype: float64

Aspen/Pitkin County Airport (ASE) had the highest average delay time

In [73]:
df_flightdata[df_flightdata["Origin"] == "ASE"]["Cancelled"].value_counts(normalize = True).mul(100).round(2)

Cancelled
False    90.45
True      9.55
Name: proportion, dtype: float64

About 9.55% of flights from ASE were cancelled.